### DISCUSSION TITLE
Possible security issue: Leaking of access token

### DISCUSSION THREAD

**Discussion by danielmarschall** [rustdesk/rustdesk#7432] [General]
I was studying the source code and found the following line which seems a bit odd:
https://github.com/rustdesk/rustdesk/blob/master/src/client.rs#L319

```
        if !key.is_empty() && !token.is_empty() {
            // mainly for the security of token
            allow_err!(secure_tcp(&mut socket, key).await);
        }
```

I understand it this way: The connection is secured by the rendezvous message "KeyExchange" to make sure that `token` and `key` are sent encrypted to the punchhole request.

However, if this `secure_tcp()` call fails, the program flow continues.

Here is a possible attack vector I have thought of (but I cannot verify it because I don't have Pro):

1. Attacker intercepts connection and drops or destroys packages which contain the "KeyExchange" rendezvous message.
2. The client gets a timeout in `secure_tcp()`, because it did not receive a `KeyExchange` message ([code](https://github.com/rustdesk/rustdesk/blob/master/src/common.rs#L1279)),
3. The client continues the program flow (without encrypted TCP channel).
4. The client [sends token and key unencrypted](https://github.com/rustdesk/rustdesk/blob/master/src/client.rs#L336)
5. The attacker can read token and key
6. The attacker can now do a lot of things with the access token and key, i.e. communicating with the API server

What do you think?

**Comment by danielmarschall** [comment]
Possible solution A: If `secure_tcp` gets a timeout (or the secure connection fails due to other problems),
then client.rs should reject the connection request. Though, there might be a risk that there connection failures when the key exchange didn't work for some reason...

Another solution B: If `secure_tcp` fails, then `key` and `token` are not sent to the punchhole request - like an "anonymous" connection. Though I think you might not want this, because if I understand correctly, creating (or maliciously forcing) such "anonymous" connections (without token/key) could bypass the license checks?

---

**_Anothing thing I have noticed:_ (Issue 2)**

Another call to `secure_tcp` is the rendezvous mediator.

I understand that `secure_tcp` does not work in the OpenSource version, because it receives a timeout due to the missing `KeyExchange` implementation at server-side.

Since the rendezvous mediator is also included in the OpenSource version, it would mean that in the OpenSource version, there will be always a timeout/delay inside the rendezvous mediator, and the connection is not encrypted.

I don't have knowledge about the rendezvous protocol, so I am not sure if the rendezvous mediator contains confidential material for the OpenSource variant. If it DOES contain secret information, then you really should encrypt it in the OpenSource version, too. (Security should not be a paid feature). If it DOES NOT contain secret information, then I think everything is fine regarding the security in this second issue.

**Comment by Kbstsn** [comment]
Is there a solution now?

**Reply by open-trade** [reply]
@danielmarschall  Before sharing your thoughts, you'd better understand [how RustDesk works](https://github.com/rustdesk/rustdesk/wiki/How-does-RustDesk-work%3F) first, which part requires encryption, which part does not. The Rendezvous protocol used in WebRTC is not encrypted either, and the TLS handshake process itself is also not encrypted. However, the actual data transmission is encrypted once the connection is established via Rendezvous protocol. Additionally, the 'access token' feature is only available in the Pro version — this feature is not present in the open-source (OSS) edition.

**Reply by danielmarschall** [reply]
It is not my job to understand the mechanics below. I lead a company and need to ensure that my customers are not in risk. This is why I asked a **_question_** more than a year ago, and did not get a reply. I just know that today every communication is encrypted, even if you visit someones hobby page, it also has HTTPS nowadays, and I know that even non-critical contents could become critical if there are specific attack vectors known. Just because some parts don't need to be encrypted, does not mean that they shall not be additionally be protected. But thanks for your explanation, if you say it is safe, then I have no other choice than trust you.

**Reply by danielmarschall** [reply]
Edit: Reading my Original Post from 2024, I still believe that there can be an attack vector if a MitM could interrupt the KeyExchange package to be sent, in order to enforce an unecrypted connection (for Pro users), and leak their token. But honestly, I don't care anymore. My company unfortunately cannot have Pro, because of the limited payment options, so I stay with OSS and don't have that problem.

**Comment by danielmarschall** [comment]
@rustdesk

Following an internal review by a colleague and myself, we have strong reason to believe that a security vulnerability exists that affects PRO users. Given the potential impact, we believe this issue warrants serious attention and may need to be reported to the appropriate authorities if left unresolved.

We are requesting an official statement from your team regarding this matter.

Our current understanding of the issue is as follows:

* PRO Users: If the `KeyExchange` package is dropped or invalidated by a Man-in-the-Middle (MitM) attacker, the token appears to be transmitted unencrypted. This would allow the MitM to intercept and read the token, effectively granting them the ability to take over the PRO account.

* OSS Users: Based on @open-trade's assertion that there is no sensitive data transmitted which "needs" protection, it may appear there is no critical risk for OSS users. However, I am not familiar with the specifics of the Rendezvous protocol and cannot verify this claim. It is worth noting that even seemingly benign data (such as an ID without key material) can pose a security risk when combined with other attack vectors.

Please confirm whether you have assessed this behavior, and if so, what your position is regarding its severity and your planned response.

**Reply by rustdesk** [reply]
https://github.com/rustdesk/rustdesk/commit/836950354bb16b55115f1f888fd7824d0d68a14e

**Reply by danielmarschall** [reply]
Thank you!
